Implement value reflection in scheme

dolio opened the pull request:
This PR is part of the implementation of value reflection in scheme, as well as many other modifications to the way things work that were implemented as part of that effort. Following is a summary of some of the changes.

- A new `unison/data-info` module is generated from unison data. It contains hash and constructor tag information and in-scheme constructor functions for various datatypes that are used in builtin functions. This depends on little other functionality (mostly `unison/data`), so it can be very early in the dependency graph, and things like `unison/primops` can use it to directly implement builtins without needing wrappers.
- The `LOAD` and `CACH` primops now check the dependency information for the objects they're called with, so that they shouldn't try to evaluate scheme code that won't actually work, and instead report the necessary additional dependencies in such cases.
- Universal compare/eq works for more runtime objects.
- `describe-value` has been expanded for better debug/trace messages covering more cases
- `unison/data` `unison/core` and `unison/boot` are now written in the racket language instead of r6rs. Mixing the two causes quite a few headaches
- Racket structs are now used for various builtin types, like term/type links. One thing this avoid is a sort of circularity that resulted from references being represented as unison data (which would have a `Reference` reference in the representation, which was a data, which had a `Reference` reference ...). But, this is also necessary for proper reflection, because otherwise it is unclear what unison type a data value is supposed to correspond to. There must be some kind of runtime tag corresponding to the type to properly implement reflection.
- One of the structs is for partial application of unison functions. A property has been specified that allows these structs to act like ordinary scheme procedures. But we can also detect them and get their partial argument list for reflection purposes.
- Some pseudo-builtins have been added to replace (un)wrappers for these structs. The pure unison code goes through serialization to mediate between the two representations, but there are much more efficient scheme functions that accomplish the same thing, so the latter are swapped in for the former in the scheme pipeline.

The serialized test cases have been updated to check that reflecting and re-serializing the deserialized test case results in the same bytes. This should exercise the new reflection machinery somewhat.

I merged in trunk at some point during development, but I seem to be out of date again. I think I also have some hard-coding to replace still in light of some of the above changes. I also need to do the share end of this and cut a new `internal` release to point to. But otherwise I think this is pretty close to working.

dolio committed:
Generate new data-info file during genlibs

dolio committed:
Add a store for known data type hashes

dolio committed:
Use data-info file in runtime value/code decoding

dolio committed:
Do dependency checking in LOAD primop

dolio committed:
Flesh out universal-compare

- Now handling byte arrays and data types

dolio committed:
Port data, core and boot modules to racket

Lets us use some racket stuff for e.g. record definitions, and cuts down
on mutable vs. immutable list nonsense.

dolio committed:
Add structure for partial app, export other structures

dolio committed:
Use partial application structure in define-unison

This should make under applied functions yield an inspectable structure
that allows us to recover their arguments easily.

dolio committed:
More internal structs, use struct-out to export

- Wrappers for code and quoted values so that they can be detected for
  reflection as distinct from the unison data types representing them.
- struct-out exports all the bindings associated with a struct at once,
  and allows using the `match` construct to do case analysis on them.

dolio committed:
Some work-in-progress code for reflection, committed for merge

dolio committed:
Merge remote-tracking branch 'origin/trunk' into topic/scheme-reflect

dolio committed:
Merge remote-tracking branch 'origin/trunk' into topic/scheme-reflect

dolio committed:
Expand termlink reps to contain reference data

dolio committed:
Add builtins for replacement in the scheme backend

These aren't actual builtins as far as the unison runtime is concerned,
but the scheme backend expects them to exist so that they can be used to
replace certain complicated, generated functions.

They are simple (un)wrapper functions for the Code and Value records. In
unison these functions are defined via serialization, but the scheme
backend can just add/remove the wrapper around the actual unison data
type representation. These are used in the implementation of
(de)serialization, so having builtins avoids a circularity.

dolio committed:
Export base32 conversions from core

dolio committed:
Various additions

- Use `match` for `describe-value`
- Fill in various new cases in `describe-value`
- Fill in code case for universal equality

dolio committed:
Adjust load/cache primops for various wrappers.

dolio committed:
Handle more builtins in describe-value

dolio committed:
Add a mechanism for associating functions to hashes

dolio committed:
Re-export some stuff from boot.ss

- The unison code has been changed to emit term and type link
  definitions as part of code generation.
  * For typelinks this allows the links to be shared among data
    constructors used in the code. Data values directly store a
    link to the type they represent.
  * The termlinks are added to a map tracking the association of
    functions to the links, for reflection purposes
- The association function is also exported from boot so that generated
  modules can declare the links for functions they define

dolio committed:
Various primops-generated adjusgments

- Add an implementation of the Value.value builtin wrapping the
  reflection function
- Adjust various tag/link references to use the new naming scheme, which
  avoids potential name conflicts with hyphenated names
- Adjust some spots that mediate between normal and chunked strings
- Remove old reference implementations of some functions
- Implement reflection for term/typelinks, code and values
- Fix data reflection: wasn't reflecting the typelink
- Remove some miscellaneous debugging artifacts that had been committed

dolio committed:
Add a value roundtrip test case

pchiusano reviewed (approved):
Awesome, Dan! Can't wait. You can go ahead and merge this once you fix conflicts and CI passes.

dolio committed:
Define Any data info comparable to auto-generated info

- Any is not a real data type, so its conventions are not actually
  defined by a unison data declaration, but by hand written builtins.
  The auto-generator of data info isn't set up to handle such
  pseudo-data right now, and because of the hand writing, it isn't
  really necessary.

dolio committed:
Remove hard-coded data info in io-handles module

dolio committed:
Merge remote-tracking branch 'origin/trunk' into topic/scheme-reflect

- Added hand-coded boolean pseudo-data to unison/data

dolio committed:
Use auto-generate optional info in indexOf

dolio committed:
Update JitInfo for new @unison/internal release
